lib/logdna/client.rb

```diff
@@ -1,164 +1,137 @@
-        @side_buffer.write(msg)
-        @side_messages.push(message_hash(msg, opts))
-        return
+    def send_request
+      @side_message_lock.synchronize do
+        @buffer.concat(@side_messages)
+        @side_messages.clear
       end
 
-      check_side_buffer
-      buffer_size = @buffer.write(msg)
-      @messages.push(message_hash(msg, opts))
-      buffer_size
-    end
+      @request.body = {
+        e: "ls",
+        ls: @buffer
+      }.to_json
 
-    def queue_to_buffer(queue=@queue)
-      next_object = queue.shift
-      write_to_buffer(next_object[:msg], next_object[:opts])
-    end
-
-    def process_buffer(buffer_size)
-      if buffer_size > @actual_byte_limit
-        @buffer_over_limit = true
-        flush()
-        @buffer_over_limit = false
-      else
-        @task = create_flush_task
+      handle_exception = lambda do |message|
+        puts message
+        @exception_flag = true
+        @side_message_lock.synchronize do
+          @side_messages.concat(@buffer)
+        end
       end
-    end
 
-    # this should be running synchronously if @buffer_over_limit i.e. called from self.buffer
-    # else asynchronously through @task
-    def flush()
-      if defined? @@request and !@@request.nil?
-        request_messages = []
-        @lock.synchronize do
-          request_messages = @messages
-          @buffer.truncate(0)
-          @messages = []
+      begin
+        @response = Net::HTTP.start(
+          @uri.hostname,
+          @uri.port,
+          use_ssl: @uri.scheme == "https"
+        ) do |http|
+          http.request(@request)
         end
-        return if request_messages.empty?
-
-        real = {
-          e: 'ls',
-          ls: request_messages,
-        }.to_json
-
-        @@request.body = real
-        @response = Net::HTTP.start(@uri.hostname, @uri.port, use_ssl: @uri.scheme == 'https') do |http|
-          http.request(@@request)
+        if @response.is_a?(Net::HTTPForbidden)
+          puts "Please provide a valid ingestion key"
+        elsif !@response.is_a?(Net::HTTPSuccess)
+          handle_exception.call("The response is not successful ")
         end
+        @exception_flag = false
+      rescue SocketError
+        handle_exception.call("Network connectivity issue")
+      rescue Errno::ECONNREFUSED => e
+        handle_exception.call("The server is down. #{e.message}")
+      rescue Timeout::Error => e
+        handle_exception.call("Timeout error occurred. #{e.message}")
+      rescue
+        handle_exception.call("#{e.message}")
+      ensure
+        @buffer.clear
+      end
+    end
 
-        # don't kill @task if this was executed from self.buffer
-        # don't kill @task if there are queued messages
-        unless @buffer_over_limit || @messages.any? || @task.nil?
-          @task.shutdown
-          @task.kill
+    def flush
+      if @lock.try_lock
+        @flush_scheduled = false
+        if @buffer.empty? && @side_messages.empty?
```

Comment:
Alternatively, might want to reverse the logic so:
```
if @buffer.any? || @side_messages.any?
  send_request
end
@lock.unlock
```

That way there's just a single @lock.unlock call.